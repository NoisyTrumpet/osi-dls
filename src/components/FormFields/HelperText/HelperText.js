import React from 'react';
import PropTypes from 'prop-types';
import './HelperText.scss';
import classNames from 'classnames';
import { Typography } from '../../Typography';

const HelperText = ({ addClass, error, children }) => {
  const classes = classNames(
    'helper-text',
    addClass,
    { error },
  );

  return (
    <div className={classes} role={error ? 'alert' : undefined}>
      <Typography variant="helper">
        {children}
      </Typography>
    </div>
  );
};

HelperText.defaultProps = {};

HelperText.propTypes = {
  addClass: PropTypes.string,
  error: PropTypes.bool,
  children: PropTypes.node,
};

export default HelperText;
