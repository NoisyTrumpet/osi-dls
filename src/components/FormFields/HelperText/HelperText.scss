.helper-text {
  margin-top: 6px;
  padding-left: 10px;

  &.error .typography-helper {
    color: map-get($colors, 'util-error');
  }
}

