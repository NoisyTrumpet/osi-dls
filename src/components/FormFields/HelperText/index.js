export { default as HelperText } from './HelperText'; // eslint-disable-line import/prefer-default-export
