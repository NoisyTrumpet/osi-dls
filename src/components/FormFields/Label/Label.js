import React from 'react';
import PropTypes from 'prop-types';
import './Label.scss';
import classNames from 'classnames';
import { Typography } from '../../Typography';

const Label = ({
  fieldType,
  forId,
  display,
  hideLabel,
  isDisabled,
  color,
  required,
  addClass,
  suffixHelperText,
  children,
}) => {
  const classes = classNames(
    'input-label',
    fieldType,
    display,
    hideLabel && 'hide-label',
    addClass,
    { isDisabled },
  );

  const srtVariant = hideLabel && "srt";

  return (
    <label
      className={classes}
      htmlFor={forId}
    >
      <Typography variant={srtVariant || "field-label"} color={color} addClass={addClass}>
        {children}
      </Typography>
      {suffixHelperText && (
        <Typography variant="helper" addClass="optional">
          {suffixHelperText}
        </Typography>
      )}
      {!required && (
        <Typography variant={srtVariant || "helper"} addClass="optional">
          (Optional)
        </Typography>
      )}
    </label>
  );
};

Label.defaultProps = {
  fieldType: 'text',
  display: 'inline',
  isDisabled: false,
  color: '',
};

Label.propTypes = {
  fieldType: PropTypes.oneOf([
    'text',
    'radio',
    'checkbox',
    'textarea',
    'select',
  ]),
  forId: PropTypes.string,
  display: PropTypes.oneOf([
    'inline',
    'inline-block',
    'block',
  ]),
  hideLabel: PropTypes.bool,
  isDisabled: PropTypes.bool,
  color: PropTypes.string,
  suffixHelperText: PropTypes.string, // helper text that is appended to label
  required: PropTypes.bool,
  addClass: PropTypes.string,
  children: PropTypes.node,
};

export default Label;
