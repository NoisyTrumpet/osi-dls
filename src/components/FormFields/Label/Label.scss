.input-label {
  color: map-get($colors, 'gray5');
  padding-left: $space-8;
  order: 2;

  &.block {
    display: block;
  }

  &.inline-block {
    display: inline-block;
  }

  &.text {
    position: absolute;
    top: rem(-11px);
    left: rem(2px);
    transform: translate(0, 50%);

    .optional {
      font-size: rem(12px);
      padding-left: rem(2px);
      padding-right: 0px;
      position: absolute;
      background: #ffffff;
    }
    
    .typography-field-label {
      background: map-get($colors, 'brand-white');
      padding: 0 $space-4;
    }
  }
}
