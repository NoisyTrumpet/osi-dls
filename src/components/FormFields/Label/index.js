export { default as Label } from './Label'; // eslint-disable-line import/prefer-default-export
