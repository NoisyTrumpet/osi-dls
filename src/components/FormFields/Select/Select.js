import React, { useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import './Select.scss';
import classNames from 'classnames';

const Select = ({
  defaultOption,
  domId,
  error,
  isDisabled,
  name,
  options,
  onChange,
  onBlur,
  value,
  variant,
}) => {
  const classes = classNames('input-field', { error });
  const myRef = useRef(null);
  const [charCount, setCharCount] = useState(0); // Character count for selected option

  const resizeComponent = () => {
    if (myRef.current) {
      if (myRef.current.options.selectedIndex === 0) {
        setCharCount(
          defaultOption ? defaultOption.length : options[0].label.length,
        );
      } else {
        const offset = defaultOption ? 1 : 0;
        setCharCount(
          options
            ? options[myRef.current.options.selectedIndex - offset].label.length
            : 0,
        );
      }
    }
  };

  const handleChange = (e) => {
    resizeComponent();
    if (onChange) {
      onChange(e);
    }
  };

  useEffect(() => {
    resizeComponent();
  }, [options]);

  // Here we are calculating the length of a selected option,
  // multiplying it by about a character's pixel width, 7,
  // and adding some additional spacing based on the variant provided.
  const style =
    variant === 'text'
      ? {
        width: `${charCount / 2 + (1 - 0.05 * charCount)}rem`,
      }
      : {};

  return (
    <select
      name={name}
      className={classes}
      disabled={isDisabled}
      id={domId}
      aria-required="true"
      aria-invalid={error}
      onChange={handleChange}
      onBlur={onBlur}
      value={value || defaultOption}
      style={style}
      ref={myRef}
    >
      {defaultOption && (
        <option key="default" value={defaultOption} default disabled>
          {defaultOption}
        </option>
      )}
      {options &&
        options.map((option) => (
          <option
            key={option.value}
            value={option.value}
            disabled={option.disabled}
          >
            {option.label}
          </option>
        ))}
    </select>
  );
};

Select.defaultProps = {
  error: false,
  isDisabled: false,
  onChange: null,
  onBlur: null,
};

Select.propTypes = {
  name: PropTypes.string,
  error: PropTypes.bool,
  isDisabled: PropTypes.bool,
  defaultOption: PropTypes.string,
  value: PropTypes.string,
  domId: PropTypes.string,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.string,
      disabled: PropTypes.bool,
    }),
  ),
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  variant: PropTypes.oneOf(['default', 'text', 'font']),
};

export default Select;
