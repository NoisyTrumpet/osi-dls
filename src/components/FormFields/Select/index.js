export { default as Select } from './Select'; // eslint-disable-line import/prefer-default-export
