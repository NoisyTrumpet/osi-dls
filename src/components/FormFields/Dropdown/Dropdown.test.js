import React from 'react';
import { mount } from 'enzyme'; import { Formik, Form } from 'formik';
import toJson from 'enzyme-to-json';
import { Dropdown } from './';
import { Label } from '../Label';
import { HelperText } from '../HelperText';
import { Select } from '../Select';

const fontOptions = [
  {
    value: 'allMyBest',
    label: 'All My Best'
  },
  {
    value: 'buhBye',
    label: 'Buh Bye'
  }
];

describe('<Dropdown />', () => {
  it('Should render Default', () => {
    const wrapper = mount(
      <Formik initialValues={{ dropdown: '' }}>
        <Form>
          <Dropdown
            label="label"
            name="dropdown"
            domId="exampleDropdown"
            helperText="Optional helper text"
          />
        </Form>
      </Formik>,
    );

    expect(wrapper.find(Label).length).toBe(1);
    expect(wrapper.find(HelperText).length).toBe(1);
    expect(wrapper.find(Select).length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should render Font Dropdown', () => {
    const wrapper = mount(
      <Formik initialValues={{ dropdown: '' }}>
        <Form>
          <Dropdown
            label="Font"
            name="myInputName"
            domId="myDomId"
            variant="font"
            helperText="Optional helper text"
            options={fontOptions}
            required
          />
        </Form>
      </Formik>,
    );

    expect(wrapper.find(Label).length).toBe(1);
    expect(wrapper.find(HelperText).length).toBe(1);
    expect(wrapper.find(Select).length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
