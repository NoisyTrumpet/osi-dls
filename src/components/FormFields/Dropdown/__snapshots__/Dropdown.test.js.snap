// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> Should render Default 1`] = `
<Formik
  initialValues={
    Object {
      "dropdown": "",
    }
  }
>
  <Form>
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <Dropdown
        color="gray4"
        domId="exampleDropdown"
        helperText="Optional helper text"
        label="label"
        name="dropdown"
        variant="default"
      >
        <div
          className="input-field-container input-field-select has-icon icon-right variant-default"
        >
          <Select
            domId="exampleDropdown"
            error={false}
            isDisabled={false}
            name="dropdown"
            onBlur={[Function]}
            onChange={[Function]}
            value=""
            variant="default"
          >
            <select
              aria-invalid={false}
              aria-required="true"
              className="input-field"
              disabled={false}
              id="exampleDropdown"
              name="dropdown"
              onBlur={[Function]}
              onChange={[Function]}
              style={Object {}}
            />
          </Select>
          <Label
            color="gray4"
            display="block"
            fieldType="text"
            forId="exampleDropdown"
            isDisabled={false}
          >
            <label
              className="input-label text block"
              htmlFor="exampleDropdown"
            >
              <Typography
                color="gray4"
                variant="field-label"
              >
                <span
                  className="typography typography-field-label color-gray4"
                >
                  label
                </span>
              </Typography>
              <Typography
                addClass="optional"
                variant="helper"
              >
                <span
                  className="typography typography-helper optional"
                >
                  (Optional)
                </span>
              </Typography>
            </label>
          </Label>
          <Icon
            badgeColor="primary"
            name="arrows-norgiedown-bold"
            size={8}
            testId="icon"
          >
            <span
              aria-hidden="true"
              className="icon icon-arrows-norgiedown-bold icon-8"
              data-testid="icon"
            />
          </Icon>
          <HelperText>
            <div
              className="helper-text"
            >
              <Typography
                variant="helper"
              >
                <span
                  className="typography typography-helper"
                >
                  Optional helper text
                </span>
              </Typography>
            </div>
          </HelperText>
        </div>
      </Dropdown>
    </form>
  </Form>
</Formik>
`;

exports[`<Dropdown /> Should render Font Dropdown 1`] = `
<Formik
  initialValues={
    Object {
      "dropdown": "",
    }
  }
>
  <Form>
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <Dropdown
        color="gray4"
        domId="myDomId"
        helperText="Optional helper text"
        label="Font"
        name="myInputName"
        options={
          Array [
            Object {
              "label": "All My Best",
              "value": "allMyBest",
            },
            Object {
              "label": "Buh Bye",
              "value": "buhBye",
            },
          ]
        }
        required={true}
        variant="font"
      >
        <div
          className="input-field-container input-field-select has-icon icon-right variant-font font-select"
        >
          <div
            className="font-dropdown"
          >
            <InlineSVG
              aria-hidden="true"
              className="font-dropdown"
              description="Font select option"
              focus={false}
              focusable="false"
              path=""
              preProcessor={null}
              role=""
              title="Font"
            >
              <InlineSVG
                aria-hidden="true"
                cacheRequests={true}
                className="font-dropdown"
                description="Font select option"
                focus="false"
                focusable="false"
                preProcessor={null}
                role=""
                src=""
                title="Font"
                uniquifyIDs={false}
              />
            </InlineSVG>
          </div>
          <Select
            domId="myDomId"
            error={false}
            isDisabled={false}
            name="myInputName"
            onBlur={[Function]}
            onChange={[Function]}
            options={
              Array [
                Object {
                  "label": "All My Best",
                  "value": "allMyBest",
                },
                Object {
                  "label": "Buh Bye",
                  "value": "buhBye",
                },
              ]
            }
            variant="font"
          >
            <select
              aria-invalid={false}
              aria-required="true"
              className="input-field"
              disabled={false}
              id="myDomId"
              name="myInputName"
              onBlur={[Function]}
              onChange={[Function]}
              style={Object {}}
            >
              <option
                key="allMyBest"
                value="allMyBest"
              >
                All My Best
              </option>
              <option
                key="buhBye"
                value="buhBye"
              >
                Buh Bye
              </option>
            </select>
          </Select>
          <Label
            color="gray4"
            display="block"
            fieldType="text"
            forId="myDomId"
            isDisabled={false}
            required={true}
          >
            <label
              className="input-label text block"
              htmlFor="myDomId"
            >
              <Typography
                color="gray4"
                variant="field-label"
              >
                <span
                  className="typography typography-field-label color-gray4"
                >
                  Font
                </span>
              </Typography>
            </label>
          </Label>
          <Icon
            badgeColor="primary"
            name="arrows-norgiedown-bold"
            size={8}
            testId="icon"
          >
            <span
              aria-hidden="true"
              className="icon icon-arrows-norgiedown-bold icon-8"
              data-testid="icon"
            />
          </Icon>
          <HelperText>
            <div
              className="helper-text"
            >
              <Typography
                variant="helper"
              >
                <span
                  className="typography typography-helper"
                >
                  Optional helper text
                </span>
              </Typography>
            </div>
          </HelperText>
        </div>
      </Dropdown>
    </form>
  </Form>
</Formik>
`;
